Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +20,44 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return ampHoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if(current_ > 0)
+    {
+        return false;
+    }
+    else
+    {
+        return true;
+    }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+   QTime time(0,0,0,0);
+   int timeLeft;
+   if(isCharging())
+   {
+       timeLeft = (totalAmpHoursUsed() / current_) * HOUR_TO_MILLISECONDS;
+   }
+   else
+   {
+       timeLeft = ((BATTERY_AMP_HOUR_CAPACITY  - totalAmpHoursUsed()) / current_) * HOUR_TO_MILLISECONDS;
+   }
+   return time.addMSecs(timeLeft);
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
+    previousCurrent_ = current_;
+    previousTime_ = time_;
+    current_ = batteryData.current;
+    voltage_ = batteryData.voltage;
+    time_ = batteryData.time;
+    averageCurrent_ = (previousCurrent_ + current_) / 2;
+    changeInTime_ = time_.msecsTo(previousTime_) / HOUR_TO_MILLISECONDS;
+    ampHoursUsed_ += averageCurrent_ * changeInTime_;
```

Comment:
This function could probably use some empty lines to space out areas of thought, but it's ok without it too.